osu.Game.Rulesets.Osu/Mods/OsuModBubbles.cs

```diff
@@ -0,0 +1,261 @@
+        public override Type[] IncompatibleMods => new[] { typeof(OsuModBarrelRoll), typeof(OsuModMagnetised), typeof(OsuModRepel) };
+
+        private PlayfieldAdjustmentContainer adjustmentContainer = null!;
+        private BubbleContainer bubbleContainer = null!;
+
+        private readonly Bindable<int> currentCombo = new BindableInt();
+
+        private float maxSize;
+        private float bubbleRadius;
+        private double bubbleFade;
+
+        public ScoreRank AdjustRank(ScoreRank rank, double accuracy) => rank;
+
+        public void ApplyToScoreProcessor(ScoreProcessor scoreProcessor)
+        {
+            currentCombo.BindTo(scoreProcessor.Combo);
+            currentCombo.BindValueChanged(combo =>
+                maxSize = Math.Min(1.75f, (float)(1.25 + 0.005 * combo.NewValue)), true);
+        }
+
+        public void ApplyToDrawableRuleset(DrawableRuleset<OsuHitObject> drawableRuleset)
+        {
+            // Multiplying by 2 results in an initial size that is too large, hence 1.85 has been chosen
+            bubbleRadius = (float)(drawableRuleset.Beatmap.HitObjects.OfType<HitCircle>().First().Radius * 1.85f);
+            bubbleFade = drawableRuleset.Beatmap.HitObjects.OfType<HitCircle>().First().TimePreempt * 2;
+
+            // We want to hide the judgements since they are obscured by the BubbleDrawable (due to layering)
+            drawableRuleset.Playfield.DisplayJudgements.Value = false;
+
+            adjustmentContainer = drawableRuleset.CreatePlayfieldAdjustmentContainer();
+
+            adjustmentContainer.Add(bubbleContainer = new BubbleContainer());
+            drawableRuleset.KeyBindingInputManager.Add(adjustmentContainer);
+        }
+
+        protected override void ApplyIncreasedVisibilityState(DrawableHitObject hitObject, ArmedState state) => applyBubbleState(hitObject);
+
+        protected override void ApplyNormalVisibilityState(DrawableHitObject hitObject, ArmedState state) => applyBubbleState(hitObject);
+
+        private void applyBubbleState(DrawableHitObject drawableObject)
+        {
+            if (drawableObject is DrawableSlider slider)
+            {
+                slider.Body.OnSkinChanged += () => applySliderState(slider);
+                applySliderState(slider);
+            }
+
+            if (drawableObject is not DrawableOsuHitObject drawableOsuObject || !drawableObject.Judged) return;
+
+            switch (drawableOsuObject)
+            {
+                case DrawableSlider:
+                case DrawableSpinnerTick:
+                    break;
+
+                default:
+                    addBubbleForObject(drawableOsuObject);
+                    break;
+            }
+        }
+
+        private void applySliderState(DrawableSlider slider) =>
+            ((PlaySliderBody)slider.Body.Drawable).BorderColour = slider.AccentColour.Value;
+
+        private void addBubbleForObject(DrawableOsuHitObject hitObject)
```

Comment:
I think you are misunderstanding pooling, or misunderstanding `ApplyNormalVisibilityState` or something in between.

The way you have this written, every time a hit object's armed state changes, you're adding a new `LifetimeEntry`. This means a single object will likely end up with more than one.

I'm not that familiar with how the lifetime entry pooling works, but I believe you'll want to make one for each hitobject at the start of gameplay, then keep a mapping to update them when state changes, or make them self-managed.

Or just go about this in a different way not using lifetimes. Like just use a normal pool and spawn a bubble when something is hit, and despawn it if unhit?

Reply:
adressed